cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt

```diff
@@ -139,12 +148,22 @@ internal class RealCache<Key : Any, Value : Any>(
         val existingEntry = cacheEntries[key]
         if (existingEntry != null) {
             // cache entry found
-            recordWrite(existingEntry, nowNanos)
+            // remove the weight of the current entry
+            totalWeight -= existingEntry.weight
+            // weigh new entry
+            val weight = weigher.weigh(key, value)
+            recordWrite(existingEntry, nowNanos, weight)
+
             existingEntry.value = value
+            existingEntry.weight = weight
```

Comment:
value and weight need to be updated atomically (e.g put them in pair or data class)